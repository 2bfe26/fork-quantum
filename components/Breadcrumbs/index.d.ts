
import React from 'react';

export interface BreadcrumbsProps {
  items: Array<{
    label: string;
    url: string;
  }>;
  theme?: {
    colors: {},
    spacing: {},
    baseFontSize: number;
  };
}

export default class Breadcrumbs extends React.Component<BreadcrumbsProps> {}
    