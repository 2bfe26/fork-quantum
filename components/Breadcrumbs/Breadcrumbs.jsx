import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { theme as defaultTheme } from '../shared';
import Icon from '../Icon';

const BreadcrumbsItems = styled.ol`
  display: flex;
  align-items: center;

  list-style: none;

  overflow: auto;
`;

const BreadcrumbItem = styled.li`
  display: flex;
  align-items: center;

  ${({
    theme: {
      baseFontSize,
      colors: { primary, secondary, neutral },
      spacing: { xsmall },
    },
  }) => css`
    font-size: ${baseFontSize * 0.75}px;

    a {
      display: flex;
      align-items: center;

      text-decoration: none;
      color: ${primary[700]};

      :hover {
        text-decoration: underline;
        color: ${secondary[700]};
      }
    }

    span {
      color: ${neutral[700]};
    }

    & + li {
      margin-left: ${xsmall}px;
    }
  `}
`;

const BreadcrumbIcon = styled(Icon).attrs(
  ({ name = 'keyboard_arrow_right' }) => ({
    name,
    size: 'small',
    'aria-hidden': true,
  }),
)`
  ${({
    theme: {
      colors: { primary },
      spacing: { xxsmall, xsmall },
    },
    name,
  }) => css`
    margin-right: ${name === 'home' ? xxsmall : xsmall}px;
    color: ${primary[700]};
  `}
`;

const ExpandButton = styled.button.attrs(() => ({
  type: 'button',
}))`
  border: none;
  background-color: transparent;

  cursor: pointer;

  &:hover {
    font-weight: 700;
  }

  ${({
    theme: {
      colors: { primary },
    },
  }) => css`
    color: ${primary[700]};
  `}
`;

const Breadcrumbs = ({ items, theme }) => {
  if (items.length < 2) {
    throw new Error('Breadcrumbs items should be greater or equal than 2.');
  }

  const [collapsed, setCollapsed] = useState(false);

  const breadcrumbsItemsEl = useRef(null);

  const lastItemIndex = items.length - 1;

  const [firstItem, ...restItems] = items;
  const lastItem = items[lastItemIndex];

  restItems.splice(lastItemIndex - 1, 1);

  useEffect(() => {
    if (items.length > 2 && window.innerWidth) {
      const { innerWidth } = window;
      const { scrollWidth } = breadcrumbsItemsEl.current;

      setCollapsed(innerWidth < scrollWidth);
    }
  }, []);

  const renderMiddleItems = () =>
    restItems.map(item => (
      <BreadcrumbItem key={item.label} theme={theme}>
        <BreadcrumbIcon theme={theme} />
        <a href={item.url} title={item.label}>
          {item.label}
        </a>
      </BreadcrumbItem>
    ));

  const renderExpandButton = () => (
    <BreadcrumbItem theme={theme}>
      <BreadcrumbIcon theme={theme} />
      <ExpandButton
        theme={theme}
        aria-label="Show all breadcrumbs"
        title="Show all breadcrumbs"
        onClick={() => setCollapsed(false)}
      >
        ...
      </ExpandButton>
    </BreadcrumbItem>
  );

  return (
    <nav>
      <BreadcrumbsItems ref={breadcrumbsItemsEl} aria-label="breadcrumb">
        <BreadcrumbItem theme={theme}>
          <a href={firstItem.url} title={firstItem.label}>
            <BreadcrumbIcon theme={theme} name="home" />
            {firstItem.label}
          </a>
        </BreadcrumbItem>

        {collapsed ? renderExpandButton() : renderMiddleItems()}

        <BreadcrumbItem theme={theme}>
          <BreadcrumbIcon theme={theme} />
          <span>{lastItem.label}</span>
        </BreadcrumbItem>
      </BreadcrumbsItems>
    </nav>
  );
};

Breadcrumbs.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      url: PropTypes.string,
    }),
  ).isRequired,
  theme: PropTypes.shape({
    colors: PropTypes.object,
  }),
};

Breadcrumbs.defaultProps = {
  theme: defaultTheme,
};

export default Breadcrumbs;
