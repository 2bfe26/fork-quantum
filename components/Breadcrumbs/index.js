// Generated with scripts/create-component.js
import Breadcrumbs from './Breadcrumbs';

export default Breadcrumbs;
