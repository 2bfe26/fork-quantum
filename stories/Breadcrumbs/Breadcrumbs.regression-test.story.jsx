// Generated with scripts/create-componnent.js
import React from 'react';

import Breadcrumbs from '../../components/Breadcrumbs';

export default {
  title: 'Breadcrumbs',
  component: Breadcrumbs,
};

const items = [
  {
    label: 'Home',
    url: 'catho.com',
  },
  {
    label: 'Quantum',
    url: 'catho.com',
  },
  {
    label: 'Components',
    url: 'catho.com',
  },
  {
    label: 'Breadcrumbs',
    url: 'catho.com',
  },
];

const Template = args => <Breadcrumbs items={items} {...args} />;

export const Default = Template.bind({});
