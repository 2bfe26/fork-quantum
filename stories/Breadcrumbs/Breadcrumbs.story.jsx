import {
  AutoPropsApi,
  TabbedView,
  Heading,
  SimpleHighlight,
  Tab,
  StoryContainer,
  Title,
} from '@catho/quantum-storybook-ui';

import Alert from '../../components/Alert';
import Breadcrumbs from '../../components/Breadcrumbs';
import defaultContent from '../../components/Breadcrumbs/exampleContents';

export default {
  title: 'Breadcrumbs',
};

const description = `Breadcrumbs component consist of a list of links that help a user visualize a page's location within the hierarchical structure of a website, and allow navigation up to any of its "ancestors".`;

export const BreadcrumbsStory = () => (
  <>
    <Heading name="Breadcrumbs">{description}</Heading>
    <TabbedView>
      <Tab title="Usage">
        <StoryContainer>
          <Title as="h2">Importing Breadcrumbs</Title>
          <SimpleHighlight>{`import { Breadcrumbs } from '@catho/quantum';`}</SimpleHighlight>
          <br />
          <Alert skin="warning" icon="info">
            The breadcrumbs will start collapsed when list items overflows
            window width
          </Alert>
        </StoryContainer>
      </Tab>

      <Tab title="API">
        <AutoPropsApi component={Breadcrumbs} />
      </Tab>

      <Tab title="Examples">
        <StoryContainer>
          <Title as="h3">Breadcrumbs default</Title>
          <SimpleHighlight>{defaultContent.code}</SimpleHighlight>
          <SimpleHighlight>
            {`<Breadcrumbs items={defaultContent} />`}
          </SimpleHighlight>
          <br />
          <Breadcrumbs items={defaultContent} />
        </StoryContainer>
      </Tab>
    </TabbedView>
  </>
);
