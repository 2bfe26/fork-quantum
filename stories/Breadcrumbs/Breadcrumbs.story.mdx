import { Meta, Canvas, Story, ArgsTable } from '@storybook/addon-docs';

import Breadcrumbs from '../../components/Breadcrumbs';

import { Header } from '../shared';

import { 
  Default
} from './Breadcrumbs.stories.jsx';

<Meta title="Breadcrumbs" component={Breadcrumbs} />

<Header title="Breadcrumbs">
  Breadcrumbs component consist of a list of links that help a user visualize a page's location within the hierarchical structure of a website, and allow navigation up to any of its "ancestors".
</Header>

- [Importing and usage](#importing)
- [API](#api)
- [Examples](#examples)

## <a id="importing"></a>Importing and usage

```js
import { Breadcrumbs } from '@catho/quantum';
```

<Canvas>
  <Story story={Default} />
</Canvas>

## <a id="api"></a>API

<ArgsTable of={Breadcrumbs} />

