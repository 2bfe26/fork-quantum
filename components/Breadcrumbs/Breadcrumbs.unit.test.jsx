// Generated with scripts/create-component.js
import React from 'react';
import { create } from 'react-test-renderer';

import Breadcrumbs from './Breadcrumbs';

describe('<Breadcrumbs />', () => {
  it('should match the snapshots', () => {
    expect(create(<Breadcrumbs />).toJSON()).toMatchSnapshot();
  });
});
