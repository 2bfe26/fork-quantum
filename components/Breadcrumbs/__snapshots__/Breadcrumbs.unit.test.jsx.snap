// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> should match the snapshots 1`] = `
Array [
  <h1>
    Hello, do domething here!
  </h1>,
  <p>
    some prop
  </p>,
]
`;
